python/fnllm/fnllm/base/services/cached.py

```diff
@@ -97,36 +100,70 @@ async def invoke(prompt: TInput, **kwargs: Unpack[LLMInput[Any, Any, Any]]):
             name = kwargs.get("name")
             bypass_cache = kwargs.get("bypass_cache", False)
             bust_cache = kwargs.get("bust_cache", False)
+            cache_metadata = kwargs.get("cache_metadata")
 
+            #
             # If we're bypassing, invoke the delegate directly
+            #
             if bypass_cache:
                 return await delegate(prompt, **kwargs)
 
-            # If we're busting the cache, skip the cache read
-            if not bust_cache:
-                cached = await self._cache.get(key)
-                if cached is not None:
-                    await self._events.on_cache_hit(key, name)
-                    output = self._cache_adapter.wrap_output(prompt, kwargs, cached)
-                    return LLMOutput(output=output, cache_hit=True)
-
-                await self._events.on_cache_miss(key, name)
+            metadata = self._get_metadata(
+                key=key,
+                data=self._cache_adapter.get_cache_input_data(prompt, kwargs),
+                metadata=cache_metadata,
+            )
 
+            #
+            # If we're busting the cache, invoke the LLM, write the result,  and don't bother checking for collisions
+            #
+            if bust_cache:
+                result = await delegate(prompt, **kwargs)
+                await self._cache.set(key, self._dump_output(result.output), metadata)
+                return result
+
+            #
+            # Check the cache before invoking the LLM
+            #
+            cached = await self._cache.get(key)
+            if cached is not None:
+                await self._events.on_cache_hit(key, name)
+                output = self._cache_adapter.wrap_output(prompt, kwargs, cached)
+                return LLMOutput(output=output, cache_hit=True)
+
+            #
+            # If we don't have a cache hit, invoke the LLM
+            #
             result = await delegate(prompt, **kwargs)
-            input_data = self._cache_adapter.get_cache_input_data(prompt, kwargs)
-            metadata = {
-                "input": input_data,
-                "key": key,
-            }
-            input_metadata = kwargs.get("cache_metadata")
-            if input_metadata:
-                metadata["custom"] = input_metadata
-
-            await self._cache.set(
-                key,
-                self._cache_adapter.dump_raw_model(result.output),
-                metadata,
-            )
+
+            #
+            # Check for inflight collisions
+            #
+            cached = await self._cache.get(key)
```

Comment:
I believe this is alright for now. But if we are in a multi-threaded scenario, this will only delay the problem. Which might still happen. If that's the case, we might need proper synchronization structured in here, to block cache read and and another for cache write.

Reply:
Yep, totally; we could think through how we could implement a filesystem-based semaphore at some point, but this closes up most of the collision window.